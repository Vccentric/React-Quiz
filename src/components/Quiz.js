import React from "react";
import PropTypes from "prop-types";

// quiz component
const Quiz = ({ name, total }) => (
  <div className="quiz">
    <div className="name">{name}</div>
    <div className="total">{total} Questions</div>
  </div>
);

Quiz.propTypes = {
  name: PropTypes.string,
  total: PropTypes.number,
};

export default Quiz;
