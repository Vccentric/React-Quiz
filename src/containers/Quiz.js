import React, { useState } from "react";
import PropTypes from "prop-types";
import Answer from "../components/Answer";
import Results from "../components/Results";

// container component that contains a quiz
const Quiz = ({ name, questions }) => {
  // setup state
  const [quizStats, setQuizStats] = useState({
    currentIndex: 0,
    correctAnswers: 0,
    selectedAnswer: false,
    showResults: false,
  });

  // setup variables
  const { currentIndex, correctAnswers, selectedAnswer, showResults } =
    quizStats;
  const totalQuestions = questions.length || 0;
  const question = questions?.[currentIndex]?.question;
  const answers = questions?.[currentIndex]?.answers;

  // setup elements
  const answersOptions = answers?.map((item, index) => {
    const key = index * Math.random();
    const value = item || "";
    const checkAnswer = () => {
      if (selectedAnswer) return false;
      const correctAnswer = questions?.[currentIndex]?.correctAnswer;
      setQuizStats({
        ...quizStats,
        correctAnswers:
          value === correctAnswer ? correctAnswers + 1 : correctAnswers,
        selectedAnswer: true,
      });
    };
    return (
      <li key={key}>
        <Answer value={value} checkAnswer={checkAnswer} />
      </li>
    );
  });

  return (
    <div id="quiz">
      <div id="header">
        <h1 className="quiz-name">{name}</h1>
      </div>
      <div id="body">
        {!showResults && (
          <div className="">
            <h2 className="question-index-total">
              Question {currentIndex + 1}/{totalQuestions}
            </h2>
            <p className="question">{question}</p>
            <ul>{answersOptions}</ul>
            {selectedAnswer && (
              <button
                onClick={() => {
                  const nextIndex = currentIndex + 1;
                  if (nextIndex < totalQuestions) {
                    setQuizStats({
                      ...quizStats,
                      currentIndex: nextIndex,
                      selectedAnswer: false,
                    });
                  } else {
                    setQuizStats({
                      ...quizStats,
                      showResults: true,
                    });
                  }
                }}
              >
                Next
              </button>
            )}
          </div>
        )}
        {showResults && (
          <Results
            correctAnswers={correctAnswers}
            totalQuestions={totalQuestions}
          />
        )}
      </div>
    </div>
  );
};

Quiz.propTypes = {
  name: PropTypes.string,
  questions: PropTypes.array,
};

export default Quiz;
