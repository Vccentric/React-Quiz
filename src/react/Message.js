import React from 'react';
import { hot } from 'react-hot-loader';

// message component
function Message(props) {
    const { isAnswered, isCorrect } = props;
    return (
        <div
            id="message"
            style={isAnswered ? { display: 'block' } : { display: 'none' }}
            className={isCorrect ? 'correct' : 'wrong'}
        >
            {isCorrect ? 'Correct Answer!' : 'Wrong Answer!'}
        </div>
    );
}

export default hot(module)(Message);
