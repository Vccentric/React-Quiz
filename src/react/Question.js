import React from 'react';
import { hot } from 'react-hot-loader';

// question component
function Question(props) {
    const { question } = props;
    return <p id="question">{question}</p>;
}

export default hot(module)(Question);
