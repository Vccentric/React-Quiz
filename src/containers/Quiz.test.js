import "@testing-library/jest-dom";
import renderer from "react-test-renderer";
import { render, screen } from "@testing-library/react";
import React from "react";
import Quiz from "./Quiz";

// setup variables
const selectedQuiz = {
  name: "Quiz Name",
  questions: [
    {
      question: "Question?",
      answerChoices: ["1", "2", "3"],
      correctAnswer: "1",
    },
  ],
};
const setQuizList = jest.fn();
const initState = {
  currentIndex: 0,
  selectedAnswer: null,
  showResults: false,
  totalCorrectAnswers: 0,
};

test("renders correctly - snapshot", () => {
  const tree = renderer
    .create(
      <Quiz
        selectedQuiz={selectedQuiz}
        setQuizList={setQuizList}
        localState={initState}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test("renders question", () => {
  render(
    <Quiz
      selectedQuiz={selectedQuiz}
      setQuizList={setQuizList}
      localState={initState}
    />
  );
  const element = screen.getByText("Question?");
  expect(element).toBeInTheDocument();
});

test("renders results summary", () => {
  const mockState = initState;
  mockState.showResults = true;
  render(
    <Quiz
      selectedQuiz={selectedQuiz}
      setQuizList={setQuizList}
      localState={mockState}
    />
  );
  const element = screen.getByText("Results:");
  expect(element).toBeInTheDocument();
});
