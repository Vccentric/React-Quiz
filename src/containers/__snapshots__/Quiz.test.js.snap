// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly - snapshot 1`] = `
<div
  id="quiz"
>
  <div
    className="header"
  >
    <h1
      className="title"
    >
      Quiz Name
    </h1>
  </div>
  <div
    className="body"
  >
    <div
      className="question"
    >
      <h2
        className="index-total"
      >
        Question 
        1
        /
        1
      </h2>
      <p
        className="question-text"
      >
        Question?
      </p>
      <ul>
        <li>
          <button
            className="answer"
            onClick={[Function]}
          >
            1
          </button>
        </li>
        <li>
          <button
            className="answer"
            onClick={[Function]}
          >
            2
          </button>
        </li>
        <li>
          <button
            className="answer"
            onClick={[Function]}
          >
            3
          </button>
        </li>
      </ul>
      <div
        className="actions-container"
      >
        <button
          className="action quit"
          onClick={[MockFunction]}
        >
          Quit
        </button>
        <button
          className="action"
          disabled={true}
          onClick={[Function]}
        >
          Next
        </button>
      </div>
    </div>
  </div>
</div>
`;
